include/rocksdb/cache.h

```diff
@@ -485,6 +485,9 @@ class Cache {
   // should check the Value() of each handle for nullptr
   virtual void WaitAll(std::vector<Handle*>& /*handles*/) {}
 
+  // Determines if if the secondary cache is enabled.
+  virtual bool IsSecondaryCacheEnabled() { return false; }
```

Comment:
This could be a problem if we make the secondary cache configuration dynamic, i.e allow the user to add/remove a secondary cache. In the future, we might at least want to allow the user to remove it in order to ensure a clean shutdown of the secondary cache.

Maybe we can have an explicit option in LRUCacheOptions to allow the user to request that stronger cache keys be used.

Reply:
If you want to get rid of this function (good idea I think), then I think the proposed option needs to be at the DB or CF level, because the Cache might not be LRUCache. I would suggest that the option be something like `use_persistable_cache_keys` which would mean that at least for secondary cache keys applicable to secondary cache, use keys that will not cause correctness issue across DB close+re-open, including if the DB and/or secondary cache are physically migrated.

Reply:
And if `use_persistable_cache_keys` is set to false, then we should probably disable secondary cache (in how we call into Cache object).

Reply:
> And if `use_persistable_cache_keys` is set to false, then we should probably disable secondary cache (in how we call into Cache object).

It is not straight forward for user to understand the behavior: if `use_persistable_cache_keys` is set to false, then secondary cache is not applicable. How about directly have an option in DBOption `use_secondary_cache`? @anand1976 what do you think?

Reply:
Would it solve the immediate problem to simply use the new kind of cache key when file->GetUniqueId returns empty?

Reply:
I believe the hypothetical problem of a file unique id collision (DB close; files moved around; DB re-open) leading to a bad cache key collision is similar to the existing problem of a file unique id collision between a deleted file with stale cache entries and a new file. We know we need to solve the latter by completely migrating away from file->GetUniqueId. So IMHO the situation is not substantially worse in the interim by introducing more file unique id collision possibilities (on FileSystems providing it) through a persistent secondary cache.

Reply:
Should work for secondary cache if we just replace NewId() with the DbSessionId+cur_file_num